### PR TITLE
add queue stats functionality

### PR DESCRIPTION
**Pull request by harshsin**
Reviewer: @rlane

### CODE REVIEW

**Comment by rlane** [comment]
This is quadratic with the number of ports. You should only need to do one iteration over the cache.

**Comment by harshsin** [comment]
Pushed another commit addressing your comment. Please take a look.

File: modules/OVSDriver/module/src/vport.c [hunk 1]
```diff
@@ -716,6 +770,66 @@ indigo_port_queue_stats_get(
     of_queue_stats_request_xid_get(queue_stats_request, &xid);
     of_queue_stats_reply_xid_set(queue_stats_reply, xid);
 
+    of_port_no_t req_of_port_num;
+    of_queue_stats_request_port_no_get(queue_stats_request, &req_of_port_num);
+    bool dump_all_ports = req_of_port_num == OF_PORT_DEST_ALL_BY_VERSION(queue_stats_request->version);
```
Comment (rlane) [inline modules/OVSDriver/module/src/vport.c]: The OF 1.4 spec has this as OFPP_ANY.

File: modules/OVSDriver/module/src/vport.c [hunk 2]
```diff
@@ -716,6 +770,66 @@ indigo_port_queue_stats_get(
...
+        /* Skip the root qdisc */
+        if (queue_id == TC_H_ROOT) continue;
+
+
+        struct ind_ovs_port *port = ind_ovs_port_lookup_by_ifindex(rtnl_tc_get_ifindex(TC_CAST(qdisc)));
+        AIM_ASSERT(port != NULL);
+        of_port_no_t port_no = port->dp_port_no;
+
+        if ((dump_all_ports && (dump_all_queues || req_queue_id == queue_id)) ||
+            (req_of_port_num == port_no && dump_all_queues)) {
```
Comment (rlane) [inline modules/OVSDriver/module/src/vport.c]: These conditionals should be refactored.

File: modules/OVSDriver/module/src/vport.c [hunk 3]
```diff
@@ -701,7 +718,44 @@ indigo_port_queue_config_get(
...
+}
+
+static void
+queue_stats_fill(of_queue_stats_entry_t *list, struct nl_object *qdisc,
+                 of_port_no_t port_no, uint32_t queue_id)
+{
+    of_queue_stats_entry_t entry[1];
+    of_queue_stats_entry_init(entry, list->version, -1, 1);
+    if (of_list_queue_stats_entry_append_bind(list, entry) < 0) {
+        AIM_DIE("unexpected error appending to queue_stats");
```
Comment (rlane) [inline modules/OVSDriver/module/src/vport.c]: How many queues would it take to fill up a 64K OpenFlow message?

File: modules/OVSDriver/module/src/vport.c [hunk 4]
```diff
@@ -716,6 +770,66 @@ indigo_port_queue_stats_get(
...
+
+    struct nl_object *qdisc;
+    for (qdisc = nl_cache_get_first(all_qdiscs); qdisc; qdisc = nl_cache_get_next(qdisc)) {
+        uint32_t queue_id = qdisc_get_queue_id(qdisc);
+        /* Skip the root qdisc */
+        if (queue_id == TC_H_ROOT) continue;
+
+
+        struct ind_ovs_port *port = ind_ovs_port_lookup_by_ifindex(rtnl_tc_get_ifindex(TC_CAST(qdisc)));
+        AIM_ASSERT(port != NULL);
```
Comment (rlane) [inline modules/OVSDriver/module/src/vport.c]: It's very possible that there are qdiscs on interfaces not attached to IVS.

**Comment by rlane** [comment]
Instead of getting a full dump each time, create the cache in ind_ovs_port_init using nl_cache_mngr_add.

File: modules/OVSDriver/module/src/vport.c [hunk 5]
```diff
@@ -701,7 +718,44 @@ indigo_port_queue_config_get(
...
+qdisc_get_queue_id(struct nl_object *qdisc)
+{
+    uint32_t parent = rtnl_tc_get_parent(TC_CAST(qdisc));
+    if (parent == TC_H_ROOT) {
+        return TC_H_ROOT;
+    }
+
+    uint32_t minor = TC_H_MIN(parent);
+    if (minor > 0) {
+        --minor;
```
Comment (rlane) [inline modules/OVSDriver/module/src/vport.c]: What's this for?

**Comment by rlane** [comment]
Finished round of review.

**Comment by harshsin** [comment]
Thanks for taking a look. Pushed commits addressing your comments.

**Comment by rlane** [comment]
ACK

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com//job/ivs_pull_req/19/
Test PASSed.

**Pull request merged**